ocaml-lsp-server/src/custom_requests/req_merlin_call_compatible.ml

```diff
@@ -0,0 +1,180 @@
+      false
+
+  let command_of_yojson params =
+    match List.assoc_opt "command" params with
+    | Some (`String command_name) -> Some command_name
+    | _ ->
+      (* If the parameter is incorrectly formatted or missing, we refuse to build
+         the parameter, [command] is mandatory. *)
+      None
+
+  let uri_of_yojson params =
+    match List.assoc_opt "uri" params with
+    | Some uri -> Some (Uri.t_of_yojson uri)
+    | _ ->
+      (* If the parameter is incorrectly formatted or missing, we refuse to build
+         the parameter, [uri] is mandatory. *)
+      None
+
+  let stringish_of_yojson =
+    (* The function is relatively optimistic and attempts to treat literal data as
+       strings of characters. *)
+    function
+    | `String s -> Some s
+    | `Bool b -> Some (string_of_bool b)
+    | `Float f -> Some (string_of_float f)
+    | `Int i -> Some (string_of_int i)
+    | `Intlit i -> Some i
+    | _ -> None
+
+  let args_of_yojson_list args =
+    let open Option.O in
+    let+ args =
+      List.fold_left
+        ~f:(fun acc x ->
+          let* acc in
+          let+ x = stringish_of_yojson x in
+          x :: acc)
+        ~init:(Some [])
+        args
+    in
+    List.rev args
+
+  let args_of_yojson_assoc args =
+    let open Option.O in
+    let+ args =
+      List.fold_left
+        ~f:(fun acc (key, value) ->
+          let key = "-" ^ key in
+          let* acc in
+          let+ x = stringish_of_yojson value in
+          x :: key :: acc)
+        ~init:(Some [])
+        args
+    in
+    List.rev args
+
+  let args_of_yojson params =
+    match List.assoc_opt "args" params with
+    | Some (`List args) -> args_of_yojson_list args
+    | Some (`Assoc args) -> args_of_yojson_assoc args
+    | _ ->
+      (* If args is not a list or is absent, it should fail. *)
+      None
+
+  let t_of_yojson = function
+    | `Assoc params ->
+      let result_as_sexp = as_sexp_of_yojson params in
+      let open Option.O in
+      let* command = command_of_yojson params in
+      let* args = args_of_yojson params in
+      let* uri = uri_of_yojson params in
+      Some { result_as_sexp; command; args; uri }
+    | _ -> None
+end
+
+let raise_invalid_params ?data ~message () =
+  let open Jsonrpc.Response.Error in
+  raise @@ make ?data ~code:Code.InvalidParams ~message ()
+
+let from_structured_json_exn = function
+  | None -> raise_invalid_params ~message:"Expected params but received none" ()
+  | Some params -> (
+    match Request_params.t_of_yojson params with
+    | Some params -> params
+    | None ->
+      let data =
+        `Assoc
+          [ ("expectedParams", Request_params.expected)
+          ; ("receivedParams", (params :> Json.t))
+          ]
+      in
+      raise_invalid_params ~data ~message:"Unexpected params format" ())
+
+let with_pipeline state uri specs raw_args cmd_args f =
+  match Document_store.get_opt state.State.store uri with
+  | Some doc ->
+    let merlin = Document.merlin_exn doc in
```

Comment:
I think you should not rely on `merlin_exn` here but match on the possible cases as is done here for example:

https://github.com/ocaml/ocaml-lsp/blob/master/ocaml-lsp-server/src/ocaml_lsp_server.ml#L172-L174

Reply:
Done in https://github.com/ocaml/ocaml-lsp/pull/1265/commits/738dffdd0ea893b8fdb309bc037c851d7e4b20dd but it returns `Null` if the document is not a merlin one.